Feat: Added delete/hide blocks; Closes: #855

9921622 opened the pull request:
- [x] Add an active field to blocks (default=True)
- [x]  If trying to delete a block that has past students in it, have a page recommend making the block inactive instead of just giving a 500 error
- [x] Add the active field to the Block edit/create form __(form already has an all tag)__
- [x] Only display active=True blocks to students when they are registering for a course

tylerecouture commented on src/courses/forms.py:
```diff
@@ -22,6 +22,9 @@ def __init__(self, *args, **kwargs):
 
+        block_qs = Block.objects.filter(active=True)
+        self.fields['block'].queryset = block_qs
+
```
I think this should be moved above the code that sets a default if there is only one option.  As is, it sets the default based on all Blocks, but then you change the queryset afterwards.  So move this up, and then change Block.objects.count() to block_qs.count()

9921622 replied:
done

tylerecouture commented on src/courses/tests/test_models.py:
```diff
@@ -348,6 +348,21 @@ def test_default_object_created(self):
+        self.assertEqual(block, course_student.block)
+
+        # see if models.PROTECT is in place
+        self.assertRaises(Exception, block.delete)
```
Be more specific in which exact exception you are testing for.  Otherwise there could be a different exception that occurs  do to some other error and your test will pass when it shouldn't.

tylerecouture commented on src/courses/views.py:
```diff
@@ -261,6 +261,22 @@ class BlockDelete(NonPublicOnlyViewMixin, DeleteView):
+
+        course_student_qs = self.get_object().coursestudent_set
+        context["population"] = course_student_qs.count()
+        context['populated'] = course_student_qs.exists()
```
The mixture of single and double quotes is triggering me.  We don't actually have a standard in this project which to use when, but side by side like this means I can't even pretend that we do, lol.

9921622 replied:
fixed

tylerecouture commented on src/courses/views.py:
```diff
@@ -261,6 +261,22 @@ class BlockDelete(NonPublicOnlyViewMixin, DeleteView):
+            return self.get(request, *args, **kwargs)
+
+        return super().post(request, *args, **kwargs)
+
```
It's not obvious to me what you're doing here, can you add a code comment explaining the purpose of this?

9921622 replied:
done

tylerecouture reviewed (changes requested):
Excellent! A few tweaks requested.  I appreciate that you tool the opportunity to change the foreign key on_delete.  Thank you!

9921622 commented:
added requested changes

also added the ProtectedError to ```CourseModelTest.test_model_protection``` since its basically the same code as ```BlockModelTest.test_model_protection```

9921622 commented:
Removed the check for forced post request

Also removed the test for forced post request since I cant figure out how to test for 500 errors
